stream.go

```diff
@@ -906,8 +914,8 @@ func (a *csAttempt) recvMsg(m interface{}, payInfo *payloadInfo) (err error) {
 
 func (a *csAttempt) finish(err error) {
 	a.mu.Lock()
+	defer a.mu.Unlock()
```

Comment:
I'm not sure how much this affects things, but the reason we were not using `defer` here is for performance on the critical RPC path:

https://github.com/golang/go/issues/14939

Please revert these changes since they do not appear essential to the PR.

Reply:
It seems marginal at best given that this only happens at the end of a conn stream, and I don't see a benchmark verifying that it is noticeable, and the risk of forgetting it in an early return is significant, but your call; done.

Reply:
:man_shrugging: for one usage, you are right that it's minor, but it added up to something significant in our benchmarks since we had several other functions with defers on the RPC path.

Best to just avoid `defer` on any critical path until golang/go#14939 is fixed.